SI-6623 -Yrepl-use-magic-imports avoids nesting $iw wrappers

retronym opened the pull request:
Rather than adding a wrapper object for each import in the session history,
just use a single wrapper preceded by the imports which have been
interspersed with a magic import to bump context depth.

Code is still ordinarily wrapped in a `$read` object.

This is a step toward 6623-like transparency.

`retronym` takes the blame for this innovation.
`adriaanm` collaborated in its commission.
`somsnytt` batted clean-up.

The following stress test now completes; before it exhausted memory after 70 iterations.
```
for i in {1..1000}; do printf 'object bar { val foo = '$i'}; \nimport bar.foo\n foo\n val foo = ""\n foo\n'; done | ./build/pack/bin/scala -Yrepl-use-magic-imports
```

retronym committed:
Refactor Context creation

  - Make `ImportContext` a nested class
  - Avoid the constructor of the superclass `Context` needing to
    call methods on the unitialized subclass ImportContext by
    computing `isRootContext` / `depth` externally in the
    factory method.

retronym committed:
SI-6623 -Yrepl-use-magic-imports avoids nesting $iw wrappers

Rather than adding a wrapper object for each import in the session
history, just use a single wrapper preceded by the imports which
have been interspersed with a magic import to bump context depth.

Code is still ordinarily wrapped in a `$read` object.

This is a step toward 6623-like transparency.

`retronym` takes the blame for this innovation.
`adriaanm` collaborated in its commission.
`somsnytt` batted clean-up.

retronym committed:
Enable magic imports unconditionally to survey test results

To be reverted.

lrytz commented:
Nice idea! Maybe give it a test run in combination with `-Yrepl-class-based`?

som-snytt commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -237,6 +237,7 @@ trait ScalaSettings extends AbsScalaSettings
   val YmacroFresh     = BooleanSetting    ("-Ymacro-global-fresh-names", "Should fresh names in macros be unique across all compilation units")
   val Yreplsync       = BooleanSetting    ("-Yrepl-sync", "Do not use asynchronous code for repl startup")
   val Yreplclassbased = BooleanSetting    ("-Yrepl-class-based", "Use classes to wrap REPL snippets instead of objects")
+  val YreplMagicImport = BooleanSetting    ("-Yrepl-use-magic-imports", "In the code the wraps REPL snippes, use magic imports to rather than nesting wrapper object/classes")
```
sorry to snipe at `snippes`, which I guess is the French.

`In the code that wraps REPL snippets, use magic imports rather`

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Contexts.scala:
```diff
@@ -1100,6 +1101,16 @@ trait Contexts { self: Analyzer =>
+      tree match {
+        case Import(expr, selector :: Nil) =>
+          // Just a syntactic check to avoid forcing typechecking of imports
+          selector.name.string_==(nme.INTERPRETER_IMPORT_LEVEL_UP) && owner.enclosingTopLevelClass.isInterpreterWrapper
```
My other use case was to submit a directory of source files as a single compilation unit, with files wrapped in `LEVEL_UP` and `LEVEL_DOWN` and then concatenated in a single source, so that top-level statements would not collide. That would not be in a repl wrapper. If double brace isn't magical enough, how about `${{`?

Maybe also call it `WRAPL`.

retronym replied:
What's the user-level use case for that? Something to do with `sealed`?

som-snytt replied:
Yes, on that thread they were complaining that their files get too big when they cram everything in. That also covers companionship. Sometimes I feel like I'm crammed in with my companion.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -536,8 +536,10 @@ trait Namers extends MethodSynthesis {
+      // warn proactively if specific import loses to definition in scope,
+      // since it may result in desired implicit not imported into scope.
+      def checkNotRedundant(pos: Position, from: Name, to0: Name): Unit = {
+        def check(to: Name): Unit = {
```
Um. I guess since we're talking import magic. I know the check used to be wrong, false positive. Now I'm not sure what the user story is. Oh, this is covered by unused import. So is this check only guarding against old bugs in name binding? Including implicit shadowing?
```
object X {
  val global = 0
  import concurrent.ExecutionContext.global
  println(global)
}
```

som-snytt replied:
Yes, I see bug 2866 is due to Jason: "I have five unanswered questions"... after grepping for permanently hidden again in the check files...

retronym replied:
Yeah, I dropped your changes to this code from #5220 as I didn't understand the original code or the change. 🤷‍♂

som-snytt replied:
I'm not sure it's very meaningful, even if it had a bug-compatible meaning once.

som-snytt commented on src/repl/scala/tools/nsc/interpreter/Imports.scala:
```diff
@@ -147,77 +146,86 @@ trait Imports {
-    def addWrapper() {
-      import nme.{ INTERPRETER_IMPORT_WRAPPER => iw }
+    def addWrapperCode(): Unit = {
+      import nme.{INTERPRETER_IMPORT_WRAPPER => iw}
```
A lot has changed in four years, including spaces in braces.

som-snytt commented on src/repl/scala/tools/nsc/interpreter/Imports.scala:
```diff
@@ -147,77 +146,86 @@ trait Imports {
+        addLevelChangingImport()
+      } else {
+        addWrapperCode()
+      }
```
Lots of braces.

som-snytt reviewed (approved):
Way to "level up"!

retronym commented:
/rebuild 09063fc66b3debc30f9b339b072da1838b2a6839

retronym committed:
Run some completions tests under -Yrepl-classbased and fix completion bug

By default, the parser tries to heal incomplete source files by inserting
missing braces. Compilation will still error out, but any subsequent
parser/type errors make more sense to the user when this healing works.

The healing uses indentation to figure out the intent of the code. Wrapped
REPL snippets aren't properly indented, and in the test case I added the
healing seems counterproductive.

This commit disables it in REPL tab completion, in line with the way that
IMain parses the wrapped code for real compilation.

retronym committed:
Small improvements to REPL -Yrepl-class-based

  - add comments describing a problem I discovered with imports
    and a possible solution
  - Make strip wrappers from REPL output for the wrappers generated in this mode

retronym committed:
Enable -Yrepl-class-based to survey test failures

retronym commented:
> Nice idea! Maybe give it a test run in combination with `-Yrepl-class-based`?

I've enabled `-Yrepl-class-based` by default for a temporary commit sequence to survey the test failures.

I minimized these by:
  - fixing a bug in tab completions under this mode
  - aligning the line numbers in the two strategies to avoid benign but annoying differences in parse errors (in 2.12.x we don't have the logic to subtract the pre-amble span from the position of reported errors which makes these tests brittle with respect to the encoding)

I investigated another test failure which is a genuine bug in this mode. But fixing it without regressing Spark will take a separate change to do some dead-code elimination to avoid unneeded capture which is bad for serialization (either makes it slow or makes it fail)

Finally, we have errors in tests that define value classes and macro implementations in the REPL. These constructs need to be wrapped in an object. We need some sort of story for these use cases before we could enable class-based mode by default.

retronym committed:
Revert exploratory changes to default REPL wrappers

Revert "Enable magic imports unconditionally to survey test results"

This reverts commit b02f890fcab96d2461e55c9b0e9579efd41f5dd3.

Revert "Enable -Yrepl-class-based to survey test failures"

This reverts commit d64af1a634f105c06a8cfd8d4e363a0c8c66edc3.

retronym committed:
:javap in REPL deals better with nested modules

retronym committed:
Stabilize output in REPL tests between indy and inline lambdas

This detail changes under -Yrepl-class-based. We prefer tests that operate
under either mode.

som-snytt commented:
I was about to rebase when I realized this is on an old version of Scala. I was just doing `:javap` in power mode, which exacerbates the in-your-faceness. Congrats on bringing this back. It's like those seeds that lie dormant in the soil until a wildfire. Or something.

retronym commented:
I'll do the hand-woven forward port of this to 2.13.x.
